src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/randomized/IterativeImprovementBasedQueryOptimizer.java

```diff
@@ -0,0 +1,127 @@
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.QueryOptimizationContext;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.rewriting.RewritingRule;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.rewriting.RuleApplication;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.utils.CostEstimationUtils;
+import se.liu.ida.hefquin.engine.utils.Pair;
+
+public class IterativeImprovementBasedQueryOptimizer implements QueryOptimizer {
+	
+	protected final QueryOptimizationContext context;
+	protected final StoppingConditionForIterativeImprovement condition;
+
+	
+	public IterativeImprovementBasedQueryOptimizer (final QueryOptimizationContext ctxt, final StoppingConditionForIterativeImprovement x) {
+		assert ctxt != null;
+		assert x != null;
+		context = ctxt;
+		condition = x;
+	}
+	
+	
+	@Override
+	public PhysicalPlan optimize( final LogicalPlan initialPlan ) throws QueryOptimizationException {
+		return optimize( context.getLogicalToPhysicalPlanConverter().convert(initialPlan,false) );
+	}
+
+	public PhysicalPlan optimize( final PhysicalPlan initialPlan ) throws QueryOptimizationException {
+		// The best plan we have found so far. As we have only found one plan, it is the best one so far.
+		PhysicalPlan bestPlan = initialPlan;
+		PhysicalPlan currentPlan = bestPlan; // This variable will hold the plan which is currently being worked on.
+		
+		// Creating the completable future.
+		CostModel costModel = context.getCostModel();
+		CompletableFuture<Double> futureCost = costModel.initiateCostEstimation(bestPlan);
+
+		// Declaring other variables here, possibly giving the future some time to be realized with minimal waiting.
+		RewritingRule rwRule = null;
+		List<PhysicalPlan> neighbours;
+		
+		Double[] neighbourCosts;
+		List<Pair<PhysicalPlan,Double>> betterPlans = new ArrayList<>();
+		Pair<PhysicalPlan,Double> betterPlan = null;
+		Random rng = new Random();
+		
+		boolean improvementFound;
+		int generation = 0;
+		
+		// Get the cost.
+		// currentCost = best cost among all neighbours. For it to be a local minimum, none of its neighbours can have a lower cost.
+		Double currentCost;
+		try {
+			currentCost = futureCost.get();
+		} catch(final ExecutionException e) {
+			throw new QueryOptimizationException("CompletableFuture throws ExecutionException!", e);
+		} catch(final InterruptedException e) {
+			throw new QueryOptimizationException("CompletableFuture throws InterruptedException!", e);		
+		}
+		// bestCost = best possible cost out of everything that the algorithm has found.
+		Double initialCost = currentCost;
+		Double bestCost = currentCost;
+		
+		while(!condition.readyToStop(generation)) { // Currently only handles generation number as a stopping condition!
+			
+			// The randomized plan generator is to be used here. As a temporary measure, the initial plan is used.
+			currentPlan = initialPlan;
+			currentCost = initialCost;

```

Comment:
Given that you (re)initialize these two variables here at the begin of every iteration step of the outer loop, you should actually also declare these variables only here and not earlier. That way, it becomes more clear that you are using these variables only within the scope of such an iteration step.